How many late days are you using for this assignment? 

In [ ]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [29]:
# Load the given datasets
one = pd.read_csv('chronic_kidney_disease_categorical.csv')
two = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
print(one.to_string())
print(two.to_string())

     unique_id   al   su       rbc        pc         pcc          ba  htn    dm  cad appet   pe  ane  Target
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   yes   no  good  yes   no     ckd
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no    no   no  good   no   no  notckd
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes    no   no  good  yes   no     ckd
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no    no   no  good   no  yes     ckd
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no    no   no  good   no   no  notckd
5       114717  0.0  0.0    normal    normal  notpresent  notpresent   no    no   no  good   no   no  notckd
6       950823  4.0  0.0       NaN    normal  notpresent  notpresent   no    no   no  good   no   no     ckd
7       901528  0.0  0.0    normal    normal  notpresent  notpresent   no    no   no  good   no   no  notckd
8       433613  0.0

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
Dataset Overview

The Chronic Kidney Disease dataset consists of medical records that indicate whether a patient has chronic kidney disease (CKD) or not. The goal when working with this dataset is often to predict the presence of CKD based on various clinical features.

Features and Labels

Features: The features are the input variables that describe the characteristics of the patients. These can include demographic information, laboratory test results, and clinical measurements. They are used to help predict the label.

Label: The label is the target variable that you want to predict. In this case, it indicates whether the patient has Chronic Kidney Disease (CKD: 1) or not (non-CKD: 0).

Types of Features

In this dataset, the features can be categorized as either categorical, numerical, or both.

Categorical Features: These are variables that represent categories or groups. They are typically non-numeric and can include labels or classifications. Examples from the dataset might include:

- Sex: Male or Female

- Blood Pressure Category: Low, Normal, High

- Diabetes: Yes or No (indicating whether the patient has diabetes)

Numerical Features: These are variables that represent measurable quantities and are usually expressed as integers or decimals. 
Examples might include:

- Age: The age of the patient in years (e.g., 45, 60)

- Hemoglobin Level: The hemoglobin concentration in the blood (e.g., 13.4 g/dL)

- Creatinine Level: The level of creatinine in the blood, which is a marker of kidney function (e.g., 1.2 mg/dL)

Examples

Categorical Columns

Sex

Diabetes (Yes or No)

Smoking Status (Smoker or Non-smoker)

Numerical Columns

Age

Serum Creatinine: Reflects kidney function and can be measured in mg/dL.

Blood Pressure: Typically measured in mmHg (e.g., systolic/diastolic).

Summary

In summary, the Chronic Kidney Disease dataset includes various features that can be classified into categorical (e.g., sex, diabetes status) and numerical variables (e.g., age, blood pressure levels). The aim is to use these features to predict the presence or absence of chronic kidney disease in patients. This kind of analysis is valuable in medical diagnostics and patient management.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [30]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
if two[two.duplicated()].shape[0]:
    print(two[two.duplicated()].shape[0])

# Drop these duplicate rows
two.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
if one[one.duplicated()].shape[0]:
    print(one[one.duplicated()].shape[0])
one.drop_duplicates()


30
30


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
396,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
398,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,good,no,yes,ckd
399,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [31]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_dataset = pd.merge(one, two, on='unique_id', how='inner')
#Print the combined dataset
print(combined_dataset)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
2       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
3       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
4       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
477     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
478     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
479     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
480     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
481     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad  ...    bp    bgr    bu   sc    sod  pot  hemo   pcv     wbcc 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [32]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows_count = combined_dataset.isnull().any(axis=1).sum()
totalRows = combined_dataset.shape[0]
percentage_missing = (missing_rows_count / totalRows) * 100

# Print %
print(f"Percentage of rows that contain at least one missing value: {percentage_missing:.2f}%")
# Drop these rows from the dataset
combined_dataset_cleaned = combined_dataset.dropna()
# Print the Dataset
print(combined_dataset_cleaned)

Percentage of rows that contain at least one missing value: 58.92%
     unique_id   al   su       rbc      pc         pcc          ba htn   dm  \
5       881763  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
6       114717  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
8       901528  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
9       433613  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
10      621332  1.0  0.0  abnormal  normal  notpresent  notpresent  no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ..  ...   
468     261137  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
469     431871  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
470     431871  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
471     769512  0.0  0.0    normal  normal  notpresent  notpresent  no   no   
472     769512  0.0  0.0    normal  normal  notpresent  notprese

## Part 3.4: Sort the dataset according to the Labels

In [33]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted = combined_dataset.sort_values(by='Target')
sorted.reset_index(drop=True, inplace=True)
# Print the dataset
print(sorted)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       373525  1.0  0.0       NaN  abnormal     present  notpresent  yes   
2       546742  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
3       675800  1.0  1.0       NaN    normal  notpresent  notpresent   no   
4       805571  1.0  4.0  abnormal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
477     924216  0.0  0.0    normal    normal  notpresent  notpresent   no   
478     296998  0.0  0.0    normal    normal  notpresent  notpresent   no   
479     182180  0.0  0.0       NaN       NaN         NaN         NaN   no   
480     328705  0.0  0.0    normal    normal  notpresent  notpresent   no   
481     676393  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...     bp    bgr     bu    sc    sod  pot  hemo   pcv    wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [44]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded = pd.get_dummies(sorted, columns=categorical_columns, drop_first=True)

# Print the dataset
print(encoded)

     unique_id   al   su   age     bp    bgr     bu    sc    sod  pot  ...  \
0       224481  3.0  NaN  54.0   60.0  125.0   21.0   1.3  137.0  3.4  ...   
1       373525  1.0  0.0  62.0   60.0  100.0   31.0   1.6    NaN  NaN  ...   
2       546742  0.0  0.0  65.0  100.0   90.0   98.0   2.5    NaN  NaN  ...   
3       675800  1.0  1.0  60.0   80.0  172.0   32.0   2.7    NaN  NaN  ...   
4       805571  1.0  4.0  41.0   80.0  210.0  165.0  18.0  135.0  4.7  ...   
..         ...  ...  ...   ...    ...    ...    ...   ...    ...  ...  ...   
477     924216  0.0  0.0  65.0   60.0  109.0   39.0   1.0  144.0  3.5  ...   
478     296998  0.0  0.0  62.0   80.0  132.0   34.0   0.8  147.0  3.5  ...   
479     182180  0.0  0.0  54.0   70.0   76.0   28.0   0.6  146.0  3.5  ...   
480     328705  0.0  0.0  39.0   70.0  131.0   46.0   0.6  145.0  5.0  ...   
481     676393  0.0  0.0  63.0   70.0  113.0   25.0   0.6  146.0  4.9  ...   

     ba_present  htn_yes  dm_no  dm_yes  cad_yes  appet_poor  p

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Implications for Downstream Modeling

Equivalence of Approaches:

- For binary variables, using separate dummy columns (one-hot encoding) or a single binary column may yield equivalent outcomes in modeling for algorithms that do not assume normality, like tree-based models.

- However, using one-hot encoding allows for more flexibility if you ever need to introduce new categories or if your model can benefit from capturing relationships between different categories.

Higher Cardinality Variables:

- When dealing with categorical variables that can take more than two values (like "type of condiment" with five categories: ketchup, mustard, mayo, relish, and salsa), one-hot encoding would transform this column into five binary columns, each representing one condiment.

- If you were to use a single binary column (e.g., encoding as numeric values 0 to 4), this would remove the categorical nature of the variable and may introduce unintended ordinal relationships (where you treat mayonnaise as "higher" than ketchup, which is illogical).

Modeling Consequences:

- One-Hot Encoding: Preserves the categorical nature and avoids emitting false relationships between categories. This can enhance model accuracy, particularly in linear models.

- Single Column: May lead to an oversimplified understanding of the categorical variable and could hinder model performance because it conflates various categories as a single ordered feature.

Conclusion

Using one-hot encoding for categorical variables is generally the best practice when there are multiple unique values. It preserves the relationships and characteristics of the data more effectively for statistical and machine learning models, particularly for those sensitive to variable scales and relationships.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [35]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = combined_dataset[column].mean()
    std_dev = combined_dataset[column].std()
    lower_limit = mean - 3 * std_dev
    upper_limit = mean + 3 * std_dev
    # Filter the DataFrame to keep only the non-outliers
    combined_dataset = combined_dataset[(combined_dataset[column] >= lower_limit) & (combined_dataset[column] <= upper_limit)]

# Print the dataset
print(combined_dataset)

     unique_id   al   su     rbc      pc         pcc          ba  htn   dm  \
2       992643  0.0  0.0     NaN     NaN  notpresent  notpresent   no   no   
5       881763  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
6       114717  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
8       901528  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
9       433613  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...     ...         ...         ...  ...  ...   
471     769512  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
472     769512  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
479     856244  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes   
480     856244  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes   
481     856244  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes   

    cad  ...    bp    bgr    bu   sc    sod  pot  hemo   pcv   

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [36]:
# Normalize the all Numerical Attributes in the dataset.
scaler = MinMaxScaler()
combined_dataset[numerical_columns] = scaler.fit_transform(combined_dataset[numerical_columns])
# Print the dataset
print(combined_dataset)

     unique_id   al   su     rbc      pc         pcc          ba  htn   dm  \
2       992643  0.0  0.0     NaN     NaN  notpresent  notpresent   no   no   
5       881763  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
6       114717  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
8       901528  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
9       433613  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...     ...         ...         ...  ...  ...   
471     769512  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
472     769512  0.0  0.0  normal  normal  notpresent  notpresent   no   no   
479     856244  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes   
480     856244  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes   
481     856244  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes   

    cad  ...    bp       bgr        bu        sc       sod     

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id is an arbitrary number.

In [37]:
#Remove that column
combined_dataset.drop(columns=['unique_id'], inplace=True)
# Print the dataset
print(combined_dataset)

      al   su     rbc      pc         pcc          ba  htn   dm cad appet  \
2    0.0  0.0     NaN     NaN  notpresent  notpresent   no   no  no  good   
5    0.0  0.0  normal  normal  notpresent  notpresent   no   no  no  good   
6    0.0  0.0  normal  normal  notpresent  notpresent   no   no  no  good   
8    0.0  0.0  normal  normal  notpresent  notpresent   no   no  no  good   
9    0.0  0.0  normal  normal  notpresent  notpresent   no   no  no  good   
..   ...  ...     ...     ...         ...         ...  ...  ...  ..   ...   
471  0.0  0.0  normal  normal  notpresent  notpresent   no   no  no  good   
472  0.0  0.0  normal  normal  notpresent  notpresent   no   no  no  good   
479  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes  no  poor   
480  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes  no  poor   
481  0.0  0.0     NaN  normal  notpresent  notpresent  yes  yes  no  poor   

     ...    bp       bgr        bu        sc       sod       pot      hemo 

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [38]:
# Export the dataframe to a new csv file
combined_dataset.to_csv('cleaned_data.csv', index=False)
# Export the dataframe to a new json file
combined_dataset.to_json('cleaned_data.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [39]:
top_15_rows = combined_dataset.head(15)
print(top_15_rows)

     al   su       rbc      pc         pcc          ba  htn   dm  cad appet  \
2   0.0  0.0       NaN     NaN  notpresent  notpresent   no   no   no  good   
5   0.0  0.0    normal  normal  notpresent  notpresent   no   no   no  good   
6   0.0  0.0    normal  normal  notpresent  notpresent   no   no   no  good   
8   0.0  0.0    normal  normal  notpresent  notpresent   no   no   no  good   
9   0.0  0.0    normal  normal  notpresent  notpresent   no   no   no  good   
10  1.0  0.0  abnormal  normal  notpresent  notpresent   no  yes   no  poor   
14  0.0  0.0    normal  normal  notpresent  notpresent   no   no   no  good   
17  0.0  0.0    normal  normal  notpresent  notpresent   no   no   no  good   
18  0.0  0.0    normal  normal  notpresent  notpresent   no   no   no  good   
19  2.0  1.0       NaN  normal  notpresent  notpresent  yes   no   no  good   
20  2.0  0.0  abnormal  normal  notpresent  notpresent   no   no   no  good   
21  0.0  0.0    normal  normal  notpresent  notprese

Paste here: 

### Paste the markdown table here
| unique_id |   al |   su | rbc      | pc      | pcc        | ba         | htn | dm | cad | ... |    bp |       bgr |       bu |       sc |       sod |       pot |       hemo |       pcv |       wbcc |       rbcc |
|-----------|------|------|----------|---------|------------|------------|-----|----|-----|-----|-------|-----------|-----------|-----------|-----------|-----------|------------|-----------|------------|------------|
| 992643    |  0.0 |  0.0 | NaN      | NaN     | notpresent | notpresent  | no  | no | no  | ... |  0.25 | 0.038710  | 0.169935  | 0.060870  | 1.000000  | 0.263158  | 0.713043   | 0.942857  | 0.491228   | 0.902439   |
| 881763    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.25 | 0.016129  | 0.137255  | 0.069565  | 0.700000  | 0.263158  | 0.886957   | 1.000000  | 0.350877   | 0.926829   |
| 114717    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.50 | 0.167742  | 0.098039  | 0.034783  | 0.600000  | 0.657895  | 0.939130   | 0.628571  | 0.464912   | 0.682927   |
| 901528    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.50 | 0.154839  | 0.222222  | 0.026087  | 0.566667  | 0.263158  | 0.739130   | 0.742857  | 0.482456   | 0.560976   |
| 433613    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.50 | 0.087097  | 0.196078  | 0.017391  | 1.000000  | 0.526316  | 0.686957   | 0.828571  | 0.587719   | 0.634146   |
| 621332    |  1.0 |  0.0 | abnormal | normal  | notpresent | notpresent  | no  | yes | no  | ... |  0.50 | 0.751613  | 0.163399  | 0.078261  | 0.066667  | 0.263158  | 0.356522   | 0.457143  | 0.622807   | 0.463415   |
| 460465    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.00 | 0.125806  | 0.098039  | 0.060870  | 0.700000  | 0.578947  | 0.826087   | 0.628571  | 0.394737   | 0.682927   |
| 129654    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.50 | 0.177419  | 0.078431  | 0.069565  | 0.633333  | 0.552632  | 0.886957   | 0.685714  | 0.078947   | 0.536585   |
| 608641    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.50 | 0.193548  | 0.130719  | 0.060870  | 0.766667  | 0.657895  | 0.834783   | 0.742857  | 0.350877   | 0.512195   |
| 261025    |  2.0 |  1.0 | NaN      | normal  | notpresent | notpresent  | yes | no | yes | ... |  0.75 | 0.319355  | 0.248366  | 0.173913  | 0.600000  | 0.105263  | 0.617391   | 0.800000  | 0.631579   | 0.902439   |
| 532520    |  2.0 |  0.0 | abnormal | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.75 | 0.112903  | 0.281046  | 0.165217  | 0.533333  | 0.710526  | 0.417391   | 0.400000  | 0.587719   | 0.414634   |
| 168242    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.00 | 0.135484  | 0.222222  | 0.069565  | 0.733333  | 0.631579  | 0.713043   | 0.714286  | 0.491228   | 0.975610   |
| 536401    | yes  | NaN  | NaN      | NaN     | notpresent | notpresent  | yes | no | yes | ... |  1.00 | 0.106452  | 0.450980  | 0.426087  | 0.500000  | 1.000000  | 0.000000   | 0.000000  | 0.298246   | 0.048780   |
| 720366    |  0.0 |  0.0 | normal   | normal  | notpresent | notpresent  | no  | no | no  | ... |  0.50 | 0.012903  | 0.202614  | 0.008696  | 0.766667  | 0.526316  | 0.765217   | 0.828571  | 0.517544   | 0.780488   |
| 851448    |  2.0 |  2.0 | normal   | NaN     | notpresent | notpresent  | yes | yes | no  | ... |  0.50 | 0.225806  | 0.392157  | 0.260870  | 0.533333  | 0.447368  | 0.582609   | 0.600000  | 0.526316   | 0.439024   |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?
Reflecting on Hallucination Risks with Language Models

Caution: Hallucination Risks

When using language models like GPT for data manipulation tasks, there are several factors to consider regarding potential "hallucinations," such as:

Understanding Data Context: The model may misinterpret or lack context about the specific dataset, leading to incorrect assumptions about the data structure or types. This can lead to erroneous output.

Data Integrity: The processed or converted outputs may not accurately represent the original data due to erroneous transformations suggested by the model. This could be particularly problematic for larger datasets where patterns might be misinterpreted.

Complexity of Tasks: More complex transformations or queries might result in more significant discrepancies since the model may oversimplify or miss nuances in your data.

Mitigation Strategies
Here are some strategies to mitigate the risks of hallucinations from LLM conversions:
Review the Output: Always review the output provided by the model carefully. Cross-check the data types, format, and consistency with your original dataset.

Run Validation Checks: After converting or manipulating data, run validation checks to confirm that the output aligns with your expectations. For example, confirm dimension integrity or perform statistical checks (like mean, median).

Iterate Step-by-Step: Instead of asking for a complex transformation in one go, break it down into smaller, manageable steps and verify each step's output before proceeding.

Use Existing Documentation: Refer to the documentation of libraries like Pandas to validate the correctness of the suggested methods and approaches by the model.

Compare Against Known Queries: If possible, compare the suggested outputs from the model with established, known outputs from tried SQL queries or data manipulations to confirm accuracy.

Combine Human Expertise: Utilize your own understanding and intuition as a data analyst to challenge the model's output and enforce corrections where necessary.

By following these strategies, you can enhance the reliability of your data manipulation processes and minimize the risks of errors stemming from automated transformations.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [40]:
# Code to rename all the columns in the dataset
# Mapping of abbreviated column names to full names
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}
combined_dataset.rename(columns=column_mapping, inplace=True)
print(combined_dataset)

     albumin  sugar red blood cells pus cell pus cell clumps    bacteria  \
2        0.0    0.0             NaN      NaN      notpresent  notpresent   
5        0.0    0.0          normal   normal      notpresent  notpresent   
6        0.0    0.0          normal   normal      notpresent  notpresent   
8        0.0    0.0          normal   normal      notpresent  notpresent   
9        0.0    0.0          normal   normal      notpresent  notpresent   
..       ...    ...             ...      ...             ...         ...   
471      0.0    0.0          normal   normal      notpresent  notpresent   
472      0.0    0.0          normal   normal      notpresent  notpresent   
479      0.0    0.0             NaN   normal      notpresent  notpresent   
480      0.0    0.0             NaN   normal      notpresent  notpresent   
481      0.0    0.0             NaN   normal      notpresent  notpresent   

    hypertension diabetes mellitus coronary artery disease appetite  ...  \
2          

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [41]:
# Converted SQL to Pandas code
pandas = combined_dataset.groupby('Target').size().reset_index(name='count')

*Describe what the above code does here*
Breakdown of the Code:
`combined_dataset.groupby('Target')`:

- This part groups the DataFrame `combined_dataset` by the unique values found in the "Target" column.

- Each unique value in the "Target" column will form a separate group.

`.size()`:
- The `size()` function counts the number of entries (or rows) in each group created by the `groupby()` operation.

- For instance, if there are 10 rows in the "Target" group labeled "A", 5 in group "B", and 7 in group "C", the `size()` function will return a series with these counts.

`.reset_index(name='count')`:

- This method resets the index of the resulting Series to create a new DataFrame.


- The `name='count'` argument specifies that the size counts will be given the column name "count" in the new DataFrame.


- After this step, the resulting DataFrame will have two columns: one for the unique values of "Target" and another for the corresponding counts of these values.

Final Resulting Output:
The final output of the code will be a new DataFrame, commonly referred to as `count_data`, which will look something like this:
| Target | count |
|--------|-------|
| A | 10 |
| B | 5 |
| C | 7 |

Columns:

- Target: This column lists the unique values found in the original "Target" column.

- count: This column reflects the number of times each unique value appears in the dataset.

Summary of What the Code Does:

Overall, this code is a very effective way to summarize the data in the "Target" column by providing a count of occurrences for each unique identifier. This can be useful for understanding how different categories are represented within your dataset and is often used in exploratory data analysis to quickly gain insights into categorical data distributions.